### PR TITLE
provider/aws: Add Route 53 delegation set resource

### PR DESCRIPTION
**Pull request by radeksimko**
### Test plan (takes ~63mins to finish)

```
$ make testacc TEST=./builtin/providers/aws TESTARGS='-run=Route53' 2>/dev/null
```
```
go generate ./...
TF_ACC=1 go test ./builtin/providers/aws -v -run=Route53 -timeout 90m
=== RUN TestAccRoute53DelegationSet
--- PASS: TestAccRoute53DelegationSet (2.15s)
=== RUN TestAccRoute53DelegationSet_withZones
--- PASS: TestAccRoute53DelegationSet_withZones (135.51s)
=== RUN TestAccRoute53Record_basic
--- PASS: TestAccRoute53Record_basic (273.96s)
=== RUN TestAccRoute53Record_txtSupport
--- PASS: TestAccRoute53Record_txtSupport (263.60s)
=== RUN TestAccRoute53Record_generatesSuffix
--- PASS: TestAccRoute53Record_generatesSuffix (273.69s)
=== RUN TestAccRoute53Record_wildcard
--- PASS: TestAccRoute53Record_wildcard (386.95s)
=== RUN TestAccRoute53Record_weighted
--- PASS: TestAccRoute53Record_weighted (266.26s)
=== RUN TestAccRoute53Record_alias
--- PASS: TestAccRoute53Record_alias (254.61s)
=== RUN TestAccRoute53Record_weighted_alias
--- PASS: TestAccRoute53Record_weighted_alias (529.06s)
=== RUN TestAccRoute53Record_TypeChange
--- PASS: TestAccRoute53Record_TypeChange (405.93s)
=== RUN TestAccRoute53ZoneAssociation_basic
--- PASS: TestAccRoute53ZoneAssociation_basic (281.50s)
=== RUN TestAccRoute53ZoneAssociation_region
--- PASS: TestAccRoute53ZoneAssociation_region (271.70s)
=== RUN TestAccRoute53Zone_basic
--- PASS: TestAccRoute53Zone_basic (153.32s)
=== RUN TestAccRoute53Zone_private_basic
--- PASS: TestAccRoute53Zone_private_basic (148.25s)
=== RUN TestAccRoute53Zone_private_region
--- PASS: TestAccRoute53Zone_private_region (147.78s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/aws	3794.276s
```

See https://github.com/hashicorp/terraform/pull/2157

### CODE REVIEW

File: website/Gemfile.lock [hunk 1]
```diff
@@ -79,18 +79,18 @@ GEM
       celluloid (~> 0.16.0)
       rb-fsevent (>= 0.9.3)
       rb-inotify (>= 0.9)
-    middleman (3.3.13)
+    middleman (3.3.12)
```
Comment (pmoust) [inline website/Gemfile.lock]: `git add .`? :)

**Comment by pmoust** [comment]
Thanks!

File: website/Gemfile.lock (hunk 1 above)
Comment (radeksimko) [inline website/Gemfile.lock]: I'm not sure what's the question/issue ? :)

I have not been changing in the actual `Gemfile`, all these packages are resolved dependencies.

Comment (radeksimko) [inline website/Gemfile.lock]: I just moved this change into a separate PR - #2073 to make it easier for anyone to review.

**Comment by radeksimko** [comment]
I just fixed a small bug w/ `aws_route53_zone.delegation_set` and I'm just getting to the bottom of `name_servers` list interpolation.

Either way, I'm relatively sure that the interpolation issue is a separate one and will be solved globally (without having to touch any specific resource/field/config, so this PR is ready for review & merge.

**Comment by radeksimko** [comment]
btw. if anyone prefers `aws_route53_zone.delegation_set_id` over `aws_route53_zone.delegation_set`, then I'm okay with changing it, I'm just not sure what's the convention for this, but `_id` suffix actually makes more sense to me.

**Comment by phinze** [comment]
Yeah +1 for the `_id` suffix. Helps to steer people away from passing the whole resource in as the value for the attribute.

**Comment by radeksimko** [comment]
Hold off on merging this - I'm waiting for successful run of all R53 acceptance tests and it's failing for some reason, not sure if it's just https://github.com/hashicorp/terraform/pull/2158 or something more serious.

**Comment by radeksimko** [comment]
This is ready for review & merge, I just successfully finished running all mentioned acceptance tests.

File: builtin/providers/aws/resource_aws_route53_delegation_set.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+	callerRef := resource.UniqueId()
+	if v, ok := d.GetOk("reference_name"); ok {
+		callerRef = strings.Join([]string{
+			v.(string), "-", callerRef,
+		}, "")
+	}
+	input := &route53.CreateReusableDelegationSetInput{
+		CallerReference: aws.String(callerRef),
+	}
+	if v, ok := d.GetOk("zone_id"); ok {
```
Comment (catsby) [inline builtin/providers/aws/resource_aws_route53_delegation_set.go]: `zone_id` isn't in the schema?

File: builtin/providers/aws/resource_aws_route53_delegation_set_test.go [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/hashicorp/terraform/terraform"
+
+	"github.com/awslabs/aws-sdk-go/aws"
+	"github.com/awslabs/aws-sdk-go/service/route53"
+)
+
+func TestAccRoute53DelegationSet(t *testing.T) {
```
Comment (catsby) [inline builtin/providers/aws/resource_aws_route53_delegation_set_test.go]: and you add a `_basic` suffix here? Otherwise we can't easily run `TestAccRoute53DelegationSet` in insolation

**Comment by catsby** [comment]
2 nitpicks, otherwise looks great!

File: builtin/providers/aws/resource_aws_route53_delegation_set.go (hunk 2 above)
Comment (radeksimko) [inline builtin/providers/aws/resource_aws_route53_delegation_set.go]: Good point, I wonder how come this wasn't caught by any test... :confused:

File: builtin/providers/aws/resource_aws_route53_delegation_set_test.go (hunk 3 above)
Comment (radeksimko) [inline builtin/providers/aws/resource_aws_route53_delegation_set_test.go]: I will do.

File: builtin/providers/aws/resource_aws_route53_delegation_set.go (hunk 2 above)
Comment (radeksimko) [inline builtin/providers/aws/resource_aws_route53_delegation_set.go]: So this was actually in the original concept, but then I trashed the idea of defining `zone_id` directly from this resource as I can't see any meaningful use-case for it.

There's `aws_route53_zone.delegation_set_id` which I'm adding and I believe the pairing will be much more useful that way.

So I guess I can just remove these three lines unless you can think of a use-case.

File: builtin/providers/aws/resource_aws_route53_delegation_set_test.go (hunk 3 above)
Comment (radeksimko) [inline builtin/providers/aws/resource_aws_route53_delegation_set_test.go]: done

File: builtin/providers/aws/resource_aws_route53_delegation_set.go (hunk 2 above)
Comment (catsby) [inline builtin/providers/aws/resource_aws_route53_delegation_set.go]: > unless you can think of a use-case for it. 
I didn't see one on my first pass of the code / API docs

File: builtin/providers/aws/resource_aws_route53_delegation_set_test.go (hunk 3 above)
Comment (catsby) [inline builtin/providers/aws/resource_aws_route53_delegation_set_test.go]: :

**Comment by radeksimko** [comment]
@catsby When you have some spare time, would you give this a final review, please? :)

**Comment by catsby** [comment]
LGTM, feel free to merge :+1:
